### PR TITLE
Add parallel reading support to S3SessionManager.list_messages()

### PR DESCRIPTION
**Pull request by CrysisDeu**
## Description

This PR adds parallel reading support to `S3SessionManager.list_messages()` to improve performance when retrieving multiple messages from S3. The implementation uses `ThreadPoolExecutor` to fetch messages concurrently, significantly reducing latency for sessions with many messages.

**Key Changes:**
- Implemented parallel S3 reads using `ThreadPoolExecutor` in `list_messages()`
- Maintains message order regardless of completion order
- Handles individual message read failures gracefully without stopping the entire operation
- Added comprehensive test coverage including edge cases and mock-based verification

**Backward Compatibility:**
- Existing code continues to work without modification
- No breaking changes to the API

## Related Issues

https://github.com/strands-agents/sdk-python/issues/1164

## Type of Change

New feature

## Testing

How have you tested the change?  Verify that the changes do not break functionality or introduce warnings in consuming repositories: agents-docs, agents-tools, agents-cli

- [x] I ran `hatch run prepare`

**Testing Details:**
- Added 9 new test cases covering:
  - Default behavior (sequential reads)
  - Edge cases (few messages, many messages, pagination)
  - Mock-based verification of ThreadPoolExecutor configuration
- All existing tests continue to pass (1520 passed, 4 skipped)
- Verified message ordering is preserved
- Tested error handling for individual message read failures
- No new warnings introduced (existing warnings are pre-existing deprecation warnings)

## Checklist

- [x] I have read the CONTRIBUTING document
- [x] I have added any necessary tests that prove my fix is effective or my feature works
- [x] I have updated the documentation accordingly
- [x] I have added an appropriate example to the documentation to outline the feature, or no new docs are needed
- [x] My changes generate no new warnings
- [x] Any dependent changes have been merged and published

----

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: src/strands/session/s3_session_manager.py [hunk 1]
```diff
@@ -50,6 +51,7 @@ def __init__(
         boto_session: Optional[boto3.Session] = None,
         boto_client_config: Optional[BotocoreConfig] = None,
         region_name: Optional[str] = None,
+        max_parallel_reads: int = 1,
```
Comment (JackYPCOnline) [inline src/strands/session/s3_session_manager.py]: nit: we shouldn't use magic number

File: src/strands/session/s3_session_manager.py [hunk 2]
```diff
@@ -62,11 +64,20 @@ def __init__(
...
+                Defaults to 1 (sequential) for backward compatibility and safety.
+                Set to a higher value (e.g., 10) for better performance with many messages.
+                Can be overridden per-call via list_messages() kwargs.
             **kwargs: Additional keyword arguments for future extensibility.
         """
         self.bucket = bucket
         self.prefix = prefix
 
+        # Validate max_parallel_reads
+        if not isinstance(max_parallel_reads, int) or max_parallel_reads < 1:
```
Comment (JackYPCOnline) [inline src/strands/session/s3_session_manager.py]: I think we don't check inut in our SDK.

File: src/strands/session/s3_session_manager.py [hunk 3]
```diff
@@ -259,7 +270,24 @@ def update_message(self, session_id: str, agent_id: str, session_message: Sessio
...
-        """List messages for an agent with pagination from S3."""
+        """List messages for an agent with pagination from S3.
+
+        Args:
+            session_id: ID of the session
+            agent_id: ID of the agent
+            limit: Optional limit on number of messages to return
+            offset: Optional offset for pagination
+            **kwargs: Additional keyword arguments. Supports:
+                max_parallel_reads: Override the instance-level max_parallel_reads setting
```
Comment (JackYPCOnline) [inline src/strands/session/s3_session_manager.py]: Do we need to update our docs or somewhere to indicate it can be overwritten.

File: src/strands/session/s3_session_manager.py [hunk 4]
```diff
@@ -287,10 +315,51 @@ def list_messages(
...
-            for key in message_keys:
-                message_data = self._read_s3_object(key)
+            if not message_keys:
+                return messages
+
+            # Use ThreadPoolExecutor to fetch messages concurrently
+            # Allow per-call override of max_parallel_reads via kwargs, otherwise use instance default
+            max_parallel_reads_override = kwargs.get("max_parallel_reads")
+            if max_parallel_reads_override is not None:
+                if not isinstance(max_parallel_reads_override, int) or max_parallel_reads_override < 1:
```
Comment (JackYPCOnline) [inline src/strands/session/s3_session_manager.py]: same above, maybe we don't need validation for <1

File: src/strands/session/s3_session_manager.py (hunk 3 above)
Comment (CrysisDeu) [reply src/strands/session/s3_session_manager.py]: What's the doc updating process? Since it's a different package, how do I make sure the doc updates are only merged after code updates

File: src/strands/session/s3_session_manager.py (hunk 4 above)
Comment (CrysisDeu) [reply src/strands/session/s3_session_manager.py]: ok I can remove it, do we just let the error bubble up and let the SDK user to deal with it

File: src/strands/session/s3_session_manager.py (hunk 1 above)
Comment (CrysisDeu) [reply src/strands/session/s3_session_manager.py]: Sure I will change it do DEFAULT_SEQUENTIAL_READ_THREAD_COUNT

**Review by JackYPCOnline — APPROVED** [review]

File: src/strands/session/s3_session_manager.py [hunk 5]
```diff
@@ -287,10 +312,40 @@ def list_messages(
...
-            for key in message_keys:
-                message_data = self._read_s3_object(key)
+            if not message_keys:
+                return messages
+
+            # Use ThreadPoolExecutor to fetch messages concurrently
+            # Allow per-call override of max_parallel_reads via kwargs, otherwise use instance default
+            max_workers = min(kwargs.get("max_parallel_reads", self.max_parallel_reads), len(message_keys))
+
+            with ThreadPoolExecutor(max_workers=max_workers) as executor:
```
Comment (zastrowm) [inline src/strands/session/s3_session_manager.py]: Did we want to optimize the max_workers=1 case?  Right now it seems like we're always spinning up a thread whereas in n=1, we don't need to do that

File: src/strands/session/s3_session_manager.py [hunk 6]
```diff
@@ -287,10 +312,40 @@ def list_messages(
...
+                for future in as_completed(future_to_key):
+                    key = future_to_key[future]
+                    try:
+                        message_data = future.result()
+                        # Store result at the correct index to maintain order
+                        results[key_to_index[key]] = message_data
+                    except Exception as e:
+                        # Log error but continue processing other messages
+                        # Individual failures shouldn't stop the entire operation
+                        logger.warning("key=<%s> | failed to read message from s3", key, exc_info=e)
```
Comment (pgrayy) [inline src/strands/session/s3_session_manager.py]: If we fail to read an s3 object then we will end up with an invalid messages array. With that said, I think we should raise an exception here.

File: src/strands/session/s3_session_manager.py [hunk 7]
```diff
@@ -259,7 +268,23 @@ def update_message(self, session_id: str, agent_id: str, session_message: Sessio
...
-        """List messages for an agent with pagination from S3."""
+        """List messages for an agent with pagination from S3.
+
+        Args:
+            session_id: ID of the session
+            agent_id: ID of the agent
+            limit: Optional limit on number of messages to return
+            offset: Optional offset for pagination
+            **kwargs: Additional keyword arguments. Supports:
+                max_parallel_reads: Override the instance-level max_parallel_reads setting
```
Comment (pgrayy) [inline src/strands/session/s3_session_manager.py]: Nit: For proper formatting in the docs, this should be written as:
```txt
Args:
            session_id: ID of the session
            agent_id: ID of the agent
            limit: Optional limit on number of messages to return
            offset: Optional offset for pagination
            **kwargs: Additional keyword arguments. Supports:
                
                - max_parallel_reads: Override the instance-level max_parallel_reads setting
```
The empty line is critical so that the kwargs list doesn't bunch up all on one line.

File: src/strands/session/s3_session_manager.py (hunk 5 above)
Comment (JackYPCOnline) [reply src/strands/session/s3_session_manager.py]: Optimized.

File: src/strands/session/s3_session_manager.py (hunk 6 above)
Comment (JackYPCOnline) [reply src/strands/session/s3_session_manager.py]: Updated

File: src/strands/session/s3_session_manager.py [hunk 8]
```diff
@@ -287,10 +313,42 @@ def list_messages(
...
+
+            # Optimize for single worker case - avoid thread pool overhead
+            if max_workers == 1:
+                for key in message_keys:
+                    message_data = self._read_s3_object(key)
+                    if message_data:
+                        messages.append(SessionMessage.from_dict(message_data))
+                return messages
+
+            with ThreadPoolExecutor(max_workers=max_workers) as executor:
```
Comment (Unshure) [inline src/strands/session/s3_session_manager.py]: Instead of a pool of threads, can we instead create one thread and use an `async` function to call the read_s3_object concurrently?

We had a similar approach [here](https://github.com/strands-agents/sdk-python/pull/897) where an async function was created to make the s3 call.

This ended up [getting reverted](https://github.com/strands-agents/sdk-python/commit/92da54453ee3eadf3f32b1da1522cc3e9b05bb25) because of a bug when [calling `asyncio.run` on the main thread](https://github.com/vamgan/sdk-python/blob/2caf917d5f91620e661048c2f523addea7028928/src/strands/session/file_session_manager.py#L235). Sometimes the agent is run as part of another `asyncio.run` call and scheduled on the asyncio event_loop already, so you cannot make nested `asyncio.run` calls on the same thread.

To fix that issue, you can just create a new thread, and then schedule the asyncio calls on that new thread:
```
from .._async import run_async

...

async def _load_messages_concurrently(self, messages_dir: str, message_files: list[str]) -> list[SessionMessage]:
        """Load multiple message files concurrently using async."""
        ...
        return messages

...
def list_messages(
        self, session_id: str, agent_id: str, limit: Optional[int] = None, offset: int = 0, **kwargs: Any
    ) -> List[SessionMessage]:
  ...
return run_async(self._load_messages_concurrently(messages_dir, message_files))
```

**Review by Unshure — CHANGES_REQUESTED** [review]

File: src/strands/session/s3_session_manager.py (hunk 8 above)
Comment (JackYPCOnline) [reply src/strands/session/s3_session_manager.py]: Thanks for this detailed context. @zastrowm also mentioned ir. I think there are several reasons `ThreadPoolExecutor` makes more sense here:

1. Better parallelism: The current approach runs multiple synchronous S3 reads across different threads simultaneously, which is more efficient than single-thread async for this I/O-bound operation.

2. No event loop conflicts: The nested event loop issue only applies when using asyncio.run() with async S3 calls. We're using standard threading here, so no risk of that bug.

3. Consistent with design: Our session manager avoids async throughout to keep it simple and avoid async overhead, so we should maintain that consistency.

4. Universal compatibility: ThreadPoolExecutor works safely in both async and sync contexts without event loop complications.

Comment (Unshure) [reply src/strands/session/s3_session_manager.py]: Async with `asyncio` is the better design approach, but there is some nuance. For io bound tasks, it usually makes sense to schedule them on an async event_loop so we don't need to use up threads per io call. That way we only need 1 thread instead on n threads.

The way to fix the bug is to just spin up one thread, and create an event loop on that thread, we use this pattern in a few places throughout the SDK.

I'm happy to help debug or give suggestions if you are facing issues with an async approach.

Comment (Unshure) [reply src/strands/session/s3_session_manager.py]: Async with `asyncio` is the better design approach, but there is some nuance. For io bound tasks, it usually makes sense to schedule them on an async event_loop so we don't need to use up threads per io call. That way we only need 1 thread instead on n threads.

The way to fix the bug is to just spin up one thread, and create an event loop on that thread, we use this pattern in a few places throughout the SDK.

I'm happy to help debug or give suggestions if you are facing issues with an async approach.

Comment (Unshure) [reply src/strands/session/s3_session_manager.py]: Okay, I did a bunch more digging into this in this branch: https://github.com/Unshure/sdk-python/tree/async_session

Basically async did work, but I needed to add the `aiobotocore` library which ended up bringing in about 5MB of additional dependencies. With that in mind, im much more inclined to accept this threading approach.

Ive made a minor change to remove the `max_parallel_reads` and instead just rely on the formula baked into the ThreadPoolExecutor:
```
        if max_workers is None:
            # ThreadPoolExecutor is often used to:
            # * CPU bound task which releases GIL
            # * I/O bound task (which releases GIL, of course)
            #
            # We use process_cpu_count + 4 for both types of tasks.
            # But we limit it to 32 to avoid consuming surprisingly large resource
            # on many core machine.
            max_workers = min(32, (os.process_cpu_count() or 1) + 4)
```

File: src/strands/session/s3_session_manager.py
Comment (zastrowm) [inline src/strands/session/s3_session_manager.py:1]: Could the PR description be updated to account for any parameter changes?

Comment (Unshure) [reply src/strands/session/s3_session_manager.py:1]: Updated

**Review by JackYPCOnline — APPROVED** [review]
